source/lac/cuda_solver_direct.cu

```diff
@@ -0,0 +1,353 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/lac/cuda_solver_direct.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace CUDAWrappers
+{
+  namespace internal
+  {
+    void cusparsecsr2dense(cusparseHandle_t cusparse_handle,
+                           const SparseMatrix<float> &matrix,
+                           float *dense_matrix_dev)
+    {
+      auto cusparse_matrix = matrix.get_cusparse_matrix();
+
+      cusparseStatus_t cusparse_error_code =
+        cusparseScsr2dense(
+          cusparse_handle, matrix.m(), matrix.n(),
+          std::get<3>(cusparse_matrix), std::get<0>(cusparse_matrix),
+          std::get<2>(cusparse_matrix), std::get<1>(cusparse_matrix),
+          dense_matrix_dev, matrix.m());
+      AssertCusparse(cusparse_error_code);
```

Comment:
Just to help me understand the calling structure: This call is issued on the host but uses pointers on the device, so the library function `cusparseScsr2dense` does the operations on the host?

Reply:
Almost right :smile: The call is on the host, the pointers are on the device, and (presumably) `cusparseScsr2dense` is on the device. Maybe internally this is done on the host and then moved back to the device but there is really no way to know. To do something on the host, the pointers should be on the host, and the function would be called something like `cusparseScsr2denseHost`

Reply:
Thanks, that makes sense.